File: packages/backend/src/misc/convert-host.ts
import { URL } from "node:url";
import config from "@/config/index.js";
import { toASCII } from "punycode";
import Logger from "@/services/logger.js";
import { inspect } from "node:util";

const logger = new Logger("convert-host");

export function getFullApAccount(username: string, host: string | null) {
	return host
		? `${username}@${toPuny(host)}`
		: `${username}@${toPuny(config.host)}`;
}

export function isSelfHost(host: string) {
	if (host == null) return true;
	return toPuny(config.host) === toPuny(host);
}

export function isSameOrigin(src: unknown): boolean | null {
	if (typeof src !== "string") {
		logger.debug(`unknown origin: ${inspect(src)}`);
		return null;
	}
	try {
		const u = new URL(src);
		return u.origin === config.url;
	} catch (e) {
		logger.debug(inspect(e));
		return false;
	}
}

export function extractDbHost(uri: string) {
	const url = new URL(uri);
	return toPuny(url.hostname);
}

export function toPuny(host: string) {
	return toASCII(host.toLowerCase());
}

export function toPunyNullable(host: string | null | undefined): string | null {
	if (host == null) return null;
	return toASCII(host.toLowerCase());
}

File: packages/backend-rs/src/misc/convert_host.rs
//! This module is used in the TypeScript backend only.
// We may want to (re)implement these functions in the `federation` module
// in a Rusty way (e.g., traits of actor type) if needed.

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Idna error: {0}")]
    Idna(#[from] idna::Errors),
    #[error("Url parse error: {0}")]
    UrlParse(#[from] url::ParseError),
    #[error("Hostname is missing")]
    NoHostname,
}

#[crate::ts_export]
pub fn get_full_ap_account(username: &str, host: Option<&str>) -> Result<String, Error> {
    Ok(match host {
        Some(host) => format!("{}@{}", username, to_puny(host)?),
        None => format!("{}@{}", username, extract_host(&crate::config::CONFIG.url)?),
    })
}

#[crate::ts_export]
pub fn is_self_host(host: Option<&str>) -> Result<bool, Error> {
    Ok(match host {
        Some(host) => extract_host(&crate::config::CONFIG.url)? == to_puny(host)?,
        None => true,
    })
}

#[crate::ts_export]
pub fn is_same_origin(uri: &str) -> Result<bool, Error> {
    Ok(url::Url::parse(uri)?.origin().ascii_serialization() == crate::config::CONFIG.url)
}

#[crate::ts_export]
pub fn extract_host(uri: &str) -> Result<String, Error> {
    url::Url::parse(uri)?
        .host_str()
        .ok_or(Error::NoHostname)
        .and_then(|v| Ok(to_puny(v)?))
}

#[crate::ts_export]
pub fn to_puny(host: &str) -> Result<String, idna::Errors> {
    idna::domain_to_ascii(host)
}
